hudi-client/src/main/java/org/apache/hudi/metadata/HoodieMetadataImpl.java

```diff
@@ -0,0 +1,1064 @@
+          ValidationUtils.checkArgument(instant.isCompleted());
+          HoodieRestoreMetadata restoreMetadata = TimelineMetadataUtils.deserializeHoodieRestoreMetadata(
+              timeline.getInstantDetails(instant).get());
+          update(restoreMetadata, instant.getTimestamp());
+          break;
+        }
+        case HoodieTimeline.SAVEPOINT_ACTION: {
+          ValidationUtils.checkArgument(instant.isCompleted());
+          // Nothing to be done here
+          break;
+        }
+        default: {
+          throw new HoodieException("Unknown type of action " + instant.getAction());
+        }
+      }
+    }
+  }
+
+  /**
+   * Return an ordered list of instants which have not been synced to the Metadata Table.
+
+   * @param datasetMetaClient {@code HoodieTableMetaClient} for the dataset
+   */
+  public List<HoodieInstant> findInstantsToSync(HoodieTableMetaClient datasetMetaClient) {
+    HoodieTimeline metaTimeline = metaClient.reloadActiveTimeline();
+
+    // All instants since the last time metadata table was compacted are candidates for sync
+    Option<String> compactionTimestamp = getLatestCompactionTimestamp();
+
+    // If there has not been any compaction then the first delta commit instant should be the one at which
+    // the metadata table was created. We should not sync any instants before that creation time.
+    Option<HoodieInstant> oldestMetaInstant = Option.empty();
+    if (!compactionTimestamp.isPresent()) {
+      oldestMetaInstant = Option.fromJavaOptional(metaTimeline.getInstants()
+          .filter(i -> i.isCompleted() && i.getAction().equals(HoodieTimeline.DELTA_COMMIT_ACTION))
+          .findFirst());
+      if (oldestMetaInstant.isPresent()) {
+        // TODO: Ensure this is the instant at which we created the metadata table
+      }
+    }
+
+    String metaSyncTimestamp = compactionTimestamp.isPresent() ? compactionTimestamp.get()
+        : oldestMetaInstant.isPresent() ? oldestMetaInstant.get().getTimestamp() : "";
+
+    // Metadata table is updated when an instant is completed except for the following:
+    //  CLEAN: metadata table is updated during inflight. So for CLEAN we accept inflight actions.
+    List<HoodieInstant> datasetInstants = datasetMetaClient.getActiveTimeline().getInstants()
+        .filter(i -> i.isCompleted() || (i.getAction().equals(HoodieTimeline.CLEAN_ACTION) && i.isInflight()))
+        .filter(i -> metaSyncTimestamp.isEmpty()
+            || HoodieTimeline.compareTimestamps(i.getTimestamp(), HoodieTimeline.GREATER_THAN_OR_EQUALS,
+                metaSyncTimestamp))
+        .collect(Collectors.toList());
+
+    // Each operation on dataset leads to a delta-commit on the metadata MOR table. So find only delta-commit
+    // instants in metadata table which are after the last compaction.
+    Map<String, HoodieInstant> metadataInstantMap = metaTimeline.getInstants()
+        .filter(i -> i.isCompleted() && i.getAction().equals(HoodieTimeline.DELTA_COMMIT_ACTION)
```

Comment:
this can be rewritten using `findInstantsInRange()`

Reply:
Rewritten using findInstantsAfterOrEquals